ambry-router/src/main/java/com/github/ambry/router/TtlUpdateOperation.java

```diff
@@ -388,6 +403,61 @@ private void checkAndMaybeComplete() {
     }
   }
 
+  /**
+   * Try to replicate the blob and then retry the TtlUpdate Request
+   * @return true if it's under retry. return false if cannot do retry or retry is finished.
+   */
+  private boolean retryWithReplicateBlob() {
+    // the conditions to trigger ReplicateBlob operation:
+    // 1. at least one replica returned successful status
+    // 2. and at least one replica returned NOT_FOUND status.
+    if (!operationTracker.hasNotFound() || operationTracker.getSuccessReplica().size() == 0) {
+      return false;
+    }
+
+    List<ReplicaId> successfulReplica = operationTracker.getSuccessReplica();
+    // it's the first time we hit the 503
+    if (replicateBlobCallback == null) {
+      DataNodeId sourceDataNode = successfulReplica.get(0).getDataNodeId();
+      replicateBlobCallback = new ReplicateBlobCallback(blobId, sourceDataNode);
+
+      nonBlockingRouter.replicateBlob(blobId.getID(), this.getClass().getSimpleName(), sourceDataNode,
+          replicateBlobCallback);
+      LOGGER.trace("Start the on-demand replication {} {}.", replicateBlobCallback.getBlobId(),
+          replicateBlobCallback.getSourceDataNode());
+      return true;
+    }
+
+    ReplicateBlobCallback.RetryState state = replicateBlobCallback.getState();
```

Comment:
if we get rid of the State, we can change this to something like
```
if (replicateBlobCallback == null) { 
    // start replicateBlobCallback
} else {
    if (!replicateBlobCallback.isDone()) {
        // wait for this callback to be done
    } else {
        if (replicateBlobCallback.getException() != null) {
            // it's over
        } else {
            operationCompleted = false;
            operationException.set(null);
            operationTracker = new SimpleOperationTracker(....);
            ttlUpdateRequestInfos.clear() // Don't forget about this one.
        }
    }
}
```

Reply:
If TtlUpdate retry fails, we may end with a loop to retry again and again, right?